readers/mutation_readers.cc

```diff
@@ -1548,3 +1548,104 @@ flat_mutation_reader_v2 make_compacting_reader(flat_mutation_reader_v2 source, g
         const tombstone_gc_state& gc_state, streamed_mutation::forwarding fwd) {
     return make_flat_mutation_reader_v2<compacting_reader>(std::move(source), compaction_time, get_max_purgeable, gc_state, fwd);
 }
+
+namespace {
+
+class slicing_forwarding_reader : public flat_mutation_reader_v2::impl {
+    flat_mutation_reader_v2 _underlying;
+    std::unique_ptr<const query::partition_slice> _stored_ps;
+    const query::partition_slice* _ps;
+    partition_region _prev_region = partition_region::partition_end;
+    query::clustering_row_ranges::const_iterator _it;
+    query::clustering_row_ranges::const_iterator _end;
+
+private:
+    void set_ranges(const dht::decorated_key& key) {
+        const auto& row_ranges = _ps->row_ranges(*_schema, key.key());
+        _it = row_ranges.begin();
+        _end = row_ranges.end();
+    }
+    future<bool> next_range() {
+        if (_it == _end) {
+            return make_ready_future<bool>(false);
+        }
+        return _underlying.fast_forward_to(position_range::from_range(*_it)).then([this] () mutable {
+            ++_it;
+            return true;
+        });
+    }
+
+public:
+    explicit slicing_forwarding_reader(flat_mutation_reader_v2 rd, const query::partition_slice& ps)
+        : impl(rd.schema(), rd.permit()), _underlying(std::move(rd)), _ps(&ps)
+    {
+        if (_ps->options.contains(query::partition_slice::option::reversed)) {
+            _stored_ps = std::make_unique<const query::partition_slice>(query::legacy_reverse_slice_to_native_reverse_slice(*_schema, *_ps));
+            _ps = _stored_ps.get();
+        }
```

Comment:
Don't we lose the static row and partition tombstone if we query two different partition slices from the same rd? The first ps will show the tombstone and static row, but the second won't.

Reply:
This reader reads a single slice, I'm not sure what do you mean by query two different slices. The result will be delivered as a single partition, which will only contain the rows that intersect with the slice's clustering ranges.

Reply:
I added another comment to https://github.com/scylladb/scylladb/pull/11736/commits/1d13d3e41cff267f0db900e58090fee626ed61ba but GH decided it is not worth showing on the discussions page.

Reply:
I meant two different clustering ranges in the partition slice.

But in that case, we'd only see the static row once too.